kani-compiler/src/codegen_cprover_gotoc/reachability.rs

```diff
@@ -0,0 +1,586 @@
+///
+/// For example, if `&u8` is being converted to `&dyn Debug`, this method would return:
+/// `Some(u8, dyn Debug)`.
+///
+/// This method also handles nested cases and `std` smart pointers. E.g.:
+///
+/// Conversion between `Rc<Wrapper<T>>` into `Rc<Wrapper<dyn Debug>>` should return:
+/// `Some(T, dyn Debug)`
+///
+/// The first step of this method is to extract the pointee types. Then we need to traverse the
+/// pointee types to find the actual trait and the type implementing it.
+///
+/// TODO: Do we need to handle &Wrapper<dyn T1> to &dyn T2 or is that taken care of with super
+/// trait handling? What about CoerceUnsized trait?
+fn find_trait_conversion<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    src_ty: Ty<'tcx>,
+    dst_ty: Ty<'tcx>,
+) -> Option<(Ty<'tcx>, Ty<'tcx>)> {
+    let vtable_metadata = |mir_type: Ty<'tcx>| {
+        let (metadata, _) = mir_type.ptr_metadata_ty(tcx, |ty| ty);
+        metadata != tcx.types.unit && metadata != tcx.types.usize
+    };
+
+    trace!(?dst_ty, ?src_ty, "find_trait_conversion");
+    let dst_ty_inner = find_pointer(tcx, dst_ty);
+    let src_ty_inner = find_pointer(tcx, src_ty);
+
+    trace!(?dst_ty_inner, ?src_ty_inner, "find_trait_conversion");
+    (vtable_metadata(dst_ty_inner) && !vtable_metadata(src_ty_inner)).then(|| {
+        let param_env = ParamEnv::reveal_all();
+        tcx.struct_lockstep_tails_erasing_lifetimes(src_ty_inner, dst_ty_inner, param_env)
+    })
+}
+
+/// This function extracts the pointee type of a regular pointer and std smart pointers.
+///
+/// TODO: Refactor this to use `CustomCoerceUnsized` logic which includes custom smart pointers.
+///
+/// E.g.: For `Rc<dyn T>` where the Rc definition is:
+/// ```
+/// pub struct Rc<T: ?Sized> {
+///    ptr: NonNull<RcBox<T>>,
+///    phantom: PhantomData<RcBox<T>>,
+/// }
+///
+/// pub struct NonNull<T: ?Sized> {
+///    pointer: *const T,
+/// }
+/// ```
+///
+/// This function will return `pointer: *const T`.
+fn find_pointer<'tcx>(tcx: TyCtxt<'tcx>, typ: Ty<'tcx>) -> Ty<'tcx> {
+    struct ReceiverIter<'tcx> {
+        pub curr: Ty<'tcx>,
+        pub tcx: TyCtxt<'tcx>,
+    }
+
+    impl<'tcx> LayoutOfHelpers<'tcx> for ReceiverIter<'tcx> {
+        type LayoutOfResult = TyAndLayout<'tcx>;
+
+        #[inline]
+        fn handle_layout_err(&self, err: LayoutError, span: Span, ty: Ty<'tcx>) -> ! {
+            span_bug!(span, "failed to get layout for `{}`: {}", ty, err)
+        }
+    }
+
+    impl<'tcx> HasParamEnv<'tcx> for ReceiverIter<'tcx> {
+        fn param_env(&self) -> ParamEnv<'tcx> {
+            ParamEnv::reveal_all()
+        }
+    }
+
+    impl<'tcx> HasTyCtxt<'tcx> for ReceiverIter<'tcx> {
+        fn tcx(&self) -> TyCtxt<'tcx> {
+            self.tcx
+        }
+    }
+
+    impl<'tcx> HasDataLayout for ReceiverIter<'tcx> {
+        fn data_layout(&self) -> &TargetDataLayout {
+            self.tcx.data_layout()
+        }
+    }
+
+    impl<'tcx> Iterator for ReceiverIter<'tcx> {
```

Comment:
This level of inner impls makes my head hurt

Reply:
yeah, this code needs some polishing for sure. I'll create another issue to do that.

Reply:
I'll move these definitions to be out of the function. I also need to rewrite how we're handling unsized coercion to support custom smart pointers. I created https://github.com/model-checking/kani/pull/1683 to capture this work.